File: day17/day17.c
#include <assert.h>
#include <inttypes.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

enum
{
	/* opcodes */
	OP_ADD	= 1,		/* [c] = [a] + [b]        */
	OP_MUL	= 2,		/* [c] = [a] * [b]        */
	OP_IN	= 3,		/* [a] = <input>          */
	OP_OUT	= 4,		/* output([a])            */
	OP_JNZ	= 5,		/* if ([a] != 0) goto [b] */
	OP_JZ	= 6,		/* if ([a] == 0) goto [b] */
	OP_TLT	= 7,		/* [c] = [a] < [b]        */
	OP_TEQ	= 8,		/* [c] = [a] == [b]       */
	OP_ARB  = 9,		/* rbp += [a]             */
	OP_HALT = 99,

	/* addressing modes */
	PMODE = 0,		/* absolute index */
	IMODE = 1,		/* immediate literal */
	RMODE = 2,		/* relative index (to rbp) */

	/* state of the execution */
	INPUT_EMPTY = 0,
	OUTPUT_FULL = 1,
	HALTED = 2,
};

struct module
{
	int64_t *ram;
	size_t size;
	int64_t pc;		/* instruction/program counter */
	int64_t rbp;		/* relative base pointer */

	int64_t inq[32];
	size_t ri, wi;

	int64_t outq[32];
	size_t ro, wo;

	FILE *output;		/* echoes the input and output */
};

static void module_free(struct module *m)
{
	if (m)
	{
		free(m->ram);
		free(m);
	}
}

static struct module *module_new(void)
{
	struct module *m = calloc(1, sizeof(*m));
	if (m)
	{
		m->ram = NULL;
		m->size = 0;
		m->pc = 0;
		m->ri = m->wi = m->ro = m->wo = 0;
	}
	return m;
}

static int64_t address_of(struct module *m, int64_t pos, int mode)
{
	while ((size_t)pos >= m->size)
	{
		size_t nsize = m->size ? m->size * 2 : 1024;
		int64_t *nram = realloc(m->ram, nsize * sizeof(*nram));
		assert(nram);

		/* NOTE: realloc() doesn't guarantee that the added
		 * memory is zeroed. */
		memset(nram+m->size, 0, nsize - m->size);
		m->size = nsize;
		m->ram = nram;
	}

	switch(mode)
	{
	case IMODE: return pos;
	case PMODE: return address_of(m, m->ram[pos], IMODE);
	case RMODE: return address_of(m, m->rbp + m->ram[pos], IMODE);
	default:
		fprintf(stderr, "Unknown addressing mode: %d\n", mode);
		abort();
	}
}

static void module_load(struct module *m, const int64_t *prog, size_t psize)
{
	/* NOTE: forces a reallocation */
	address_of(m, psize, IMODE);

	/* reset the memory and copy the program */
	memcpy(m->ram, prog, psize * sizeof(m->ram[0]));
	memset(m->ram+psize, 0, (m->size-psize) * sizeof(m->ram[0]));
	m->pc = 0;
	m->rbp = 0;
	m->ri = m->wi = m->ro = m->wo = 0;
}

static int module_input_full(struct module *m)
{
	return m->ri + 32 == m->wi;
}

static void module_push_input(struct module *m, int64_t value)
{
	assert(m->wi - m->ri < 32);
	if (m->output && 0 <= value && value < 256)
	{
		putc(value, m->output);
	}
	m->inq[(m->wi++) & 31] = value;
}

static int module_output_empty(struct module *m)
{
	return m->wo == m->ro;
}

static int64_t module_pop_output(struct module *m)
{
	assert(m->wo != m->ro);
	int64_t value = m->outq[(m->ro++) & 31];
	if (m->output && 0 <= value && value < 256)
	{
		putc(value, m->output);
	}
	return value;
}

static void module_dump(struct module *m, FILE *out)
{
	m->output = out;
}

static int module_execute(struct module *m)
{
	for (;;)
	{
		int op;
		div_t d = div(m->ram[m->pc], 100);
		op = d.rem;
		d = div(d.quot, 10);
		int a_mode = d.rem;
		d = div(d.quot, 10);
		int b_mode = d.rem;
		d = div(d.quot, 10);
		int c_mode = d.rem;

		int64_t a, b, c;
		switch (op)
		{
		case OP_ADD:
			a = address_of(m, m->pc+1, a_mode);
			b = address_of(m, m->pc+2, b_mode);
			c = address_of(m, m->pc+3, c_mode);
			assert(c_mode != IMODE);
			m->ram[c] = m->ram[a] + m->ram[b];
			m->pc += 4;
			break;

		case OP_MUL:
			a = address_of(m, m->pc+1, a_mode);
			b = address_of(m, m->pc+2, b_mode);
			c = address_of(m, m->pc+3, c_mode);
			assert(c_mode != IMODE);
			m->ram[c] = m->ram[a] * m->ram[b];
			m->pc += 4;
			break;

		case OP_IN:
			a = address_of(m, m->pc+1, a_mode);
			assert(a_mode != IMODE);
			if (m->wi == m->ri)
			{
				return INPUT_EMPTY;
			}
			m->ram[a] = m->inq[(m->ri++) & 31];
			m->pc += 2;
			break;

		case OP_OUT:
			a = address_of(m, m->pc+1, a_mode);
			if (m->wo - m->ro < 32)
			{
				m->outq[(m->wo++) & 31] = m->ram[a];
			}
			else
			{
				return OUTPUT_FULL;
			}
			m->pc += 2;
			break;

		case OP_JNZ:
			a = address_of(m, m->pc+1, a_mode);
			b = address_of(m, m->pc+2, b_mode);
			if (m->ram[a] != 0)
			{
				m->pc = m->ram[b];
			}
			else
			{
				m->pc += 3;
			}
			break;

		case OP_JZ:
			a = address_of(m, m->pc+1, a_mode);
			b = address_of(m, m->pc+2, b_mode);
			if (m->ram[a] == 0)
			{
				m->pc = m->ram[b];
			}
			else
			{
				m->pc += 3;
			}
			break;

		case OP_TLT:
			a = address_of(m, m->pc+1, a_mode);
			b = address_of(m, m->pc+2, b_mode);
			c = address_of(m, m->pc+3, c_mode);
			assert(c_mode != IMODE);
			m->ram[c] = m->ram[a] < m->ram[b] ? 1 : 0;
			m->pc += 4;
			break;

		case OP_TEQ:
			a = address_of(m, m->pc+1, a_mode);
			b = address_of(m, m->pc+2, b_mode);
			c = address_of(m, m->pc+3, c_mode);
			assert(c_mode != IMODE);
			m->ram[c] = m->ram[a] == m->ram[b] ? 1 : 0;
			m->pc += 4;
			break;

		case OP_ARB:
			a = address_of(m, m->pc+1, a_mode);
			m->rbp += m->ram[a];
			m->pc += 2;
			break;

		case OP_HALT:
			return HALTED;

		default:
			fprintf(stderr, "Unknown opcode %" SCNd64 " at %" SCNd64 "\n",
				m->ram[m->pc], m->pc);
			abort();
		}
	}
}

static void module_print(struct module *m)
{
	int status;
	do
	{
		status = module_execute(m);
		while (!module_output_empty(m))
		{
			if (m->outq[m->ro & 31] > 256)
			{
				break;
			}
			module_pop_output(m);
		}
	} while (status == OUTPUT_FULL);
}

static int64_t *program_load(FILE *input, size_t *count)
{
	int64_t *array = NULL;
	size_t acount = 0;
	size_t asize = 0;

	int64_t value;
	while (fscanf(input, "%" SCNd64 ",", &value) == 1)
	{
		if (acount == asize)
		{
			size_t newsize = asize ? asize * 2 : 32;
			int64_t *newarray = realloc(array, newsize * sizeof(*newarray));
			if (!newarray)
			{
				free(array);
				return NULL;
			}
			asize = newsize;
			array = newarray;
		}
		array[acount++] = value;
	}
	*count = acount;
	return array;
}

struct map
{
	char *points;
	size_t size;
	size_t count;
	unsigned width;
	unsigned height;
	unsigned startx;
	unsigned starty;
};

static void map_free(struct map *m)
{
	if (m)
	{
		free(m->points);
		free(m);
	}
}

static struct map *map_discover(struct module *mod)
{
	struct map *m = calloc(1, sizeof(*m));
	assert(m);

	int status;
	do
	{
		int64_t old = 0;
		status = module_execute(mod);
		while (!module_output_empty(mod))
		{
			int64_t v = module_pop_output(mod);
			if (v == '\n' && old != v)
			{
				old = v;
				m->height++;
				continue;
			}
			else if (m->height == 0)
			{
				m->width++;
			}

			/* take note of the current bot position */
			if (strchr("^>v<", v))
			{
				m->startx = m->count % m->width;
				m->starty = m->count / m->width;
			}

			if (m->count == m->size)
			{
				size_t nsize = m->size ? m->size * 2 : 64;
				char *npoints = realloc(m->points, nsize);
				if (!npoints)
				{
					map_free(m);
					return NULL;
				}
				m->size = nsize;
				m->points = npoints;
			}
			m->points[m->count++] = v;
		}
	} while (status == OUTPUT_FULL);
	return m;
}

static char map_get(struct map *m, unsigned x, unsigned y)
{
	if (x < m->width && y < m->height)
	{
		return m->points[y * m->width + x];
	}
	return 0;
}

static unsigned map_alignment(struct map *m)
{
	const int dx[] = {0,  0, 1, 0, -1};
	const int dy[] = {0, -1, 0, 1,  0};
	unsigned alignment = 0;
	for (unsigned y = 0; y < m->height; y++)
	{
		for (unsigned x = 0; x < m->width; x++)
		{
			int i;
			for (i = 0; i < 5; i++)
			{
				if (map_get(m, x + dx[i], y + dy[i]) != '#')
				{
					break;
				}
			}

			if (i == 5)
			{
				alignment += x*y;
			}
		}
	}
	return alignment;
}

static char *map_path(struct map *m, size_t *cmdcount)
{
	int dx, dy;
	unsigned x = m->startx;
	unsigned y = m->starty;

	switch (map_get(m, x, y))
	{
	case '^': dx = 0; dy = -1; break;
	case '>': dx = 1; dy = 0; break;
	case 'v': dx = 0; dy = 1; break;
	case '<': dx = -1; dy = 0; break;
	default:
		abort();
	}

	char *path = NULL;
	size_t psize = 0;
	size_t pcount = 0;
	cmdcount[0] = 0;
	for (;;)
	{
		/* rotate left */
		unsigned lx = dy;
		unsigned ly = -dx;

		/* rotate right */
		unsigned rx = -dy;
		unsigned ry = dx;

		/* command */
		int cmd;
		int len;

		if (map_get(m, x + dx, y + dy) == '#')
		{
			cmd = 0;
			len = 0;
			unsigned nx, ny;
			nx = x + dx;
			ny = y + dy;
			while (map_get(m, nx, ny) == '#')
			{
				x = nx;
				y = ny;
				nx += dx;
				ny += dy;
				len++;
			}
		}
		else if (map_get(m, x + lx, y + ly) == '#')
		{
			cmd = 1;
			dx = lx;
			dy = ly;
		}
		else if (map_get(m, x + rx, y + ry) == '#')
		{
			cmd = 2;
			dx = rx;
			dy = ry;
		}
		else
		{
			break;
		}

		cmdcount[0]++;
		int required;
		switch(cmd)
		{
		case 0: required = snprintf(NULL, 0, "%d,", len); break;
		case 1:
		case 2: required = 2; break;
		}
		while (pcount+required >= psize)
		{
			size_t nsize = psize ? psize * 2 : (required + 1);
			char *npath = realloc(path, nsize);
			if (!npath)
			{
				free(path);
				return NULL;
			}
			psize = nsize;
			path = npath;
		}
		switch(cmd)
		{
		case 0: snprintf(path+pcount, psize-pcount, "%d,", len); break;
		case 1: snprintf(path+pcount, psize-pcount, "L,"); break;
		case 2: snprintf(path+pcount, psize-pcount, "R,"); break;
		}
		pcount += required;
	}

	return path;
}

static int check_equal(const char **ptr, int s, int e, int pos)
{
	const char *pp;
	const char *ss;
	for (ss = ptr[s], pp = ptr[pos];
	     ss < ptr[e];
	     ss++, pp++)
	{
		if (*ss != *pp)
		{
			return 0;
		}
	}
	return 1;
}

static int skip_equal(const char **ptrs, int s, int e, int off)
{
	if (check_equal(ptrs, s, e, off))
	{
		off += e - s;
	}
	return off;
}

struct interval
{
	const char *str;
	int len;
};

static int decompose(const char *path, struct interval *inter, size_t cmdcount)
{
	/* find the start of each parameter */
	const char **ptrs = malloc((cmdcount+1) * sizeof(*ptrs));
	assert(ptrs);
	ptrs[0] = path;
	for (size_t i = 1; i < cmdcount+1; i++)
	{
		ptrs[i] = strchr(ptrs[i-1], ',')+1;
		assert(ptrs[i] != ptrs[i-1]);
	}

	/* brute force the search */
	size_t sa = 0;
	for (size_t ea = sa+2; ea < sa+20; ea += 2)
	{
		size_t sb = ea, osb = 0;
		while (sb != osb)
		{
			osb = sb;
			sb = skip_equal(ptrs, sa, ea, sb);
		}
		for (size_t eb = sb+2; eb < sb+20; eb += 2)
		{
			size_t sc = eb, osc = 0;
			while (sc != osc)
			{
				osc = sc;
				sc = skip_equal(ptrs, sa, ea, sc);
				sc = skip_equal(ptrs, sb, eb, sc);
			}
			for (size_t ec = sc+2; ec < sc+20; ec += 2)
			{
				size_t sd = ec, osd = 0;
				while (sd != osd)
				{
					osd = sd;
					sd = skip_equal(ptrs, sa, ea, sd);
					sd = skip_equal(ptrs, sb, eb, sd);
					sd = skip_equal(ptrs, sc, ec, sd);
				}
				if (sd == cmdcount)
				{
					inter[0].str = ptrs[sa];
					inter[1].str = ptrs[sb];
					inter[2].str = ptrs[sc];
					inter[0].len = ptrs[ea] - ptrs[sa];
					inter[1].len = ptrs[eb] - ptrs[sb];
					inter[2].len = ptrs[ec] - ptrs[sc];
					free(ptrs);
					return 0;
				}
			}
		}
	}
	free(ptrs);
	return -1;
}

static int64_t program_robot(struct module *mod, struct map *m)
{
	size_t cmdcount;
	char *path = map_path(m, &cmdcount);

	struct interval intervals[3];
	if (decompose(path, intervals, cmdcount) < 0)
	{
		fprintf(stderr, "Cannot decompose the command stream\n");
		free(path);
		return -1;
	}

	/* Main routine */
	module_print(mod);
	const char *t = path;
	while (*t)
	{
		int c;
		int i;
		for (i = 0; i < 3; i++)
		{
			if (strncmp(t, intervals[i].str, intervals[i].len) == 0)
			{
				t += intervals[i].len;
				c = 'A' + i;
				break;
			}
		}
		if (i < 3)
		{
			module_push_input(mod, c);
			if (*t)
			{
				module_push_input(mod, ',');
			}
		}
	}
	module_push_input(mod, '\n');

	/* A, B, C function definitions */
	module_print(mod);
	for (int i = 0; i < 3; i++)
	{
		for (int j = 0; j < intervals[i].len-1; j++)
		{
			while (module_input_full(mod))
			{
				module_execute(mod);
			}
			module_push_input(mod, intervals[i].str[j]);
		}
		while (module_input_full(mod))
		{
			module_execute(mod);
		}
		module_push_input(mod, 10);
		module_print(mod);
	}

	/* video feed */
	module_push_input(mod, 'n');
	module_push_input(mod, '\n');
	module_print(mod);
	free(path);
	return module_pop_output(mod);
}

int main(int argc, char *argv[])
{
	if (argc < 2)
	{
		fprintf(stderr, "Usage: %s <input>\n", argv[0]);
		return -1;
	}

	FILE *input = fopen(argv[1], "rb");
	if (!input)
	{
		fprintf(stderr, "File %s not found\n", argv[1]);
		return -1;
	}

	size_t pcount;
	int64_t *program = program_load(input, &pcount);
	fclose(input);
	if (!program)
	{
		fprintf(stderr, "Cannot load the program\n");
		return -1;
	}

	struct module *mod = module_new();
#ifdef DUMP_OUTPUT
	module_dump(mod, stdout);
#else
	(void)module_dump;
#endif

	module_load(mod, program, pcount);
	struct map *m = map_discover(mod);
	printf("part1: %u\n", map_alignment(m));

	/* patch the program */
	program[0] = 2;
	module_load(mod, program, pcount);
	printf("part2: %ld\n", program_robot(mod, m));

	map_free(m);
	module_free(mod);
	free(program);
	return 0;
}

File: day17/day17.py
#!/usr/bin/env python3

from collections import deque, defaultdict

import sys

ADD  = 1                        # add [a], [b] into [c]
MUL  = 2                        # mul [a], [b] into [c]
IN   = 3                        # read stdin and store to a
OUT  = 4                        # write a to stdout
JNZ  = 5                        # jump to b if a != 0
JZ   = 6                        # jump to b if a == 0
TLT  = 7                        # c = a < b
TEQ  = 8                        # c = a == b
ARB  = 9                        # add to relative base (base pointer
HALT = 99                       # halt

PMODE = 0                       # absolute index into memory
IMODE = 1                       # immediate literal
RMODE = 2                       # relative mode: offset from rbp

WAITING = 0                     # the program is waiting for input
HALTED = 1                      # the program is halted

class Module(object):
    def __init__(self):
        self.ram = defaultdict(lambda: 0)
        self.pc = 0             # instruction pointer
        self.rbp = 0            # relative base
        self.input = deque()
        self.output = deque()
        self.snapshot = None
        self.logfile = None

    def load(self, program):
        self.ram.clear()
        for i, v in enumerate(program):
            self.ram[i] = v
        self.pc = 0
        self.rbp = 0
        self.input.clear()
        self.output.clear()

    def log(self, f):
        self.logfile = f

    def push_input(self, data):
        self.input.append(data)

    def pop_output(self):
        return self.output.popleft()

    def pipe(self, module):
        self.output = module.input

    def address_of(self, addr, mode):
        "Return the address of a location for the access mode"
        if mode == IMODE:
            return addr
        elif mode == RMODE:
            return self.rbp + self.ram[addr]
        else:
            return self.address_of(self.ram[addr], IMODE)

    def execute(self):
        while True:
            op = self.ram[self.pc] % 100
            a_mode = (self.ram[self.pc] // 100) % 10
            b_mode = (self.ram[self.pc] // 1000) % 10
            c_mode = (self.ram[self.pc] // 10000) % 10
            if op == ADD:
                a = self.address_of(self.pc+1, a_mode)
                b = self.address_of(self.pc+2, b_mode)
                c = self.address_of(self.pc+3, c_mode)

                self.ram[c] = self.ram[a] + self.ram[b]
                self.pc += 4
            elif op == MUL:
                a = self.address_of(self.pc+1, a_mode)
                b = self.address_of(self.pc+2, b_mode)
                c = self.address_of(self.pc+3, c_mode)

                self.ram[c] = self.ram[a] * self.ram[b]
                self.pc += 4
            elif op == IN:
                a = self.address_of(self.pc+1, a_mode)

                if not self.input:
                    return WAITING
                self.ram[a] = self.input.popleft()
                self.pc += 2
                if self.logfile and 0 <= self.ram[a] < 256:
                    self.logfile.write(chr(self.ram[a]))
            elif op == OUT:
                a = self.address_of(self.pc+1, a_mode)
                self.pc += 2
                self.output.append(self.ram[a])
                if self.logfile and 0 <= self.ram[a] < 256:
                    self.logfile.write(chr(self.ram[a]))
                if self.ram[a] == 10:
                    return WAITING
            elif op == JNZ:
                a = self.address_of(self.pc+1, a_mode)
                b = self.address_of(self.pc+2, b_mode)

                if self.ram[a] != 0:
                    self.pc = self.ram[b]
                else:
                    self.pc += 3
            elif op == JZ:
                a = self.address_of(self.pc+1, a_mode)
                b = self.address_of(self.pc+2, b_mode)

                if self.ram[a] == 0:
                    self.pc = self.ram[b]
                else:
                    self.pc += 3
            elif op == TLT:
                a = self.address_of(self.pc+1, a_mode)
                b = self.address_of(self.pc+2, b_mode)
                c = self.address_of(self.pc+3, c_mode)

                self.ram[c] = self.ram[a] < self.ram[b] and 1 or 0
                self.pc += 4
            elif op == TEQ:
                a = self.address_of(self.pc+1, a_mode)
                b = self.address_of(self.pc+2, b_mode)
                c = self.address_of(self.pc+3, c_mode)

                self.ram[c] = (self.ram[a] == self.ram[b]) and 1 or 0
                self.pc += 4
            elif op == ARB:
                a = self.address_of(self.pc+1, a_mode)
                self.rbp += self.ram[a]
                self.pc += 2
            elif op == HALT:
                return HALTED
            else:
                print("Unknown opcode:", op, self.pc)
                assert op, "Unknown opcode"


def xmul(a, b):
    return (a[0]*b[0]-a[1]*b[1], a[0]*b[1]+a[1]*b[0])

def add(a, b):
    return (a[0]+b[0], a[1]+b[1])

def sub(a, b):
    return (a[0]-b[0], a[1]-b[1])

class Map(object):
    directions = ((0, -1), (0, +1), (-1, 0), (+1, 0))

    def __init__(self, echo = False):
        self.points = {}
        self.distances = {}
        self.max_distance = 0
        self.startpos = None
        self.startdir = None
        self.height = None
        self.module = Module()
        if echo:
            self.module.log(sys.stdout)

    def discover(self, program):
        self.module.load(program)
        while self.module.execute() != HALTED:
            pass
        x, y = 0, 0
        while self.module.output:
            v = chr(self.module.pop_output())
            if v == "\n":
                y += 1
                x = 0
            else:
                self.points[(x, y)] = v
                x += 1
        self.height = y

    def is_intersection(self, point):
        for d in self.directions:
            np = add(point, d)
            if self.points.get(np) != "#":
                return False

        return True

    def alignment(self):
        align = 0
        for point, value in self.points.items():
            if value in "#<^>v" and self.is_intersection(point):
                align += point[0] * point[1]

            p = "^>v<".find(value)
            if p != -1:
                self.startpos = point
                self.startdir = self.directions[p]

        return align

    def path(self):
        mypath = []
        curpos = self.startpos
        curdir = self.startdir
        while True:
            command = None
            leftdir = xmul(curdir, (0, -1))
            rightdir = xmul(curdir, (0, 1))
            if self.points.get(add(curpos, curdir)) == "#":
                length = 0
                npos = add(curpos, curdir)
                while self.points.get(npos) == "#":
                    curpos = npos
                    npos = add(npos, curdir)
                    length += 1
                command = length
            elif self.points.get(add(curpos, leftdir)) == "#":
                curdir = leftdir
                command = "L"
            elif self.points.get(add(curpos, rightdir)) == "#":
                curdir = rightdir
                command = "R"
            else:
                break

            mypath.append("{}".format(command))

        return mypath

    def reprogram(self, program, command, patterns):
        self.module.load(program)
        self.module.ram[0] = 2
        self.module.execute()
        while self.module.output:
            v = self.module.pop_output()

        # NOTE: main movement routine i.e. commands
        cmd = ",".join(command)+"\n"
        for v in cmd:
            self.module.push_input(ord(v))
        self.module.execute()
        while self.module.output:
            v = self.module.output.popleft()

        # NOTE: movements functions i.e. patterns
        for p in "ABC":
            cmd = ",".join(patterns[p]) + "\n"
            for v in cmd:
                self.module.push_input(ord(v))
            self.module.execute()
            while self.module.output:
                v = self.module.pop_output()

        # NOTE: disable continuous video feed
        cmd = "n\n"
        for v in cmd:
            self.module.push_input(ord(v))

        y = 0
        while True:
            self.module.execute()
            while self.module.output:
                v = self.module.pop_output()
                if v >= 256:
                    return v

        return 0

def subequal(sample, off1, off2, width):
    for i in range(width):
        sa = sample[off1+i]
        sb = sample[off2+i]
        if sa in "ABC" or sb in "ABC" or sa != sb:
            return False

    return True

def find_longest_pattern(sample, pos):
    for width in range(20, 1, -2):
        for i in range(pos+width, len(sample)-width):
            if subequal(sample, pos, i, width):
                return width

    return None

def replace(sample, src, dst):
    newsample = []
    srclen = len(src)
    j = 0
    for s in sample:
        # not quite correct but happens to work
        if s == src[j]:
            j += 1
            if j == srclen:
                newsample.append(dst)
                j = 0
        else:
            for k in range(j):
                newsample.append(src[k])
            newsample.append(s)
            j = 0

    return newsample

def build_patterns(sample):
    patterns = {}
    name = 65
    pos = 0
    worksample = sample
    while pos < len(worksample):
        if worksample[pos] in "ABC":
            pos += 1
        else:
            size = find_longest_pattern(worksample, pos)
            assert size, "Pattern not found"
            pattern = worksample[pos:pos+size]
            patterns[chr(name)] = pattern
            worksample = replace(worksample, pattern, chr(name))
            name += 1

    return worksample, patterns

program = []
with open("input", "rt") as file:
    program = [ int(x) for x in file.read().split(",") ]

m = Map(False)
m.discover(program)
print("part1:", m.alignment())
mypath = m.path()

command, patterns = build_patterns(mypath)
print("part2:", m.reprogram(program, command, patterns))
